course/views.py

```diff
@@ -58,10 +60,9 @@ def get_common_objects(self):
                     my_instances.append(instance)
             
             for instance in user.userprofile.enrolled.all().filter(ending_time__gte=end_threshold):
```

Comment:
I think it would make sense to check the visibility in the same filter call:
`for instance in user.userprofile.enrolled.all().filter(ending_time__gte=end_threshold, visible_to_students=True):`
Then it is not necessary to run `instance in all_instances`. If the user is a teacher, then the above lines have already added the instance to `my_instances`.

What do you think?

Reply:
Yes, you are right. I also tried to think of a way to save one DB query here, but doesn't seem feasible as the third one is done via userprofile, not course instances. The queries made here using your suggestion are (in pseudocode):
SELECT COURSES WHERE user is assistant OR user is teacher OR course is visible for students AND course is ongoing
SELECT COURSES WHERE user is teacher AND course is ongoing
SELECT COURSES WHERE user is enrolled AND course is visible for students AND course is ongoing